File: frontend/src/lib/api/claims.ts
import "server-only";
import { gatewayGet, gatewayPost } from "@/lib/api/gateway";
import type { ClaimStatus } from "@/lib/status";
import type { ActivityType } from "@/lib/types";

type ApiClaimStatus =
  | "submitted"
  | "ai_review"
  | "human_review"
  | "approved"
  | "rejected"
  | "more_information_requested";

const claimStatusNames: Record<ApiClaimStatus, ClaimStatus> = {
  submitted: "submitted",
  ai_review: "under_ai_review",
  human_review: "under_human_review",
  approved: "approved",
  rejected: "rejected",
  more_information_requested: "more_information_requested",
};

const apiClaimStatuses: Partial<Record<ClaimStatus, ApiClaimStatus>> = {
  submitted: "submitted",
  under_ai_review: "ai_review",
  under_human_review: "human_review",
  approved: "approved",
  rejected: "rejected",
  more_information_requested: "more_information_requested",
};

export type ClaimPriority = "normal" | "high" | "critical";

type ApiClaimEvidence = {
  evidence_id: string;
  file_name: string;
  media_type: string;
  content_hash: string;
  page_count: number | null;
  byte_size: number;
};

type ApiClaim = {
  id: string;
  facility_id: string;
  facility_name: string;
  activity_type: ActivityType;
  vintage_year: number;
  period_start: string;
  period_end: string;
  declared_figures: Record<string, string>;
  requested_amount: string;
  computed_ceiling: string;
  capacity_basis: string;
  capacity_source: string;
  discount_factor: string;
  reference_factor_value: string;
  reference_factor_id: string;
  reference_lookup_key: string;
  status: ApiClaimStatus;
  priority: ClaimPriority;
  requires_dual_approval: boolean;
  exclusivity_attested_at: string;
  issued_amount: string | null;
  created_at: string;
  evidence: ApiClaimEvidence[];
};

export type ClaimEvidenceRecord = {
  evidenceId: string;
  fileName: string;
  mediaType: string;
  contentHash: string;
  pageCount: number | null;
  byteSize: number;
};

export type ClaimRecord = {
  id: string;
  facilityId: string;
  facilityName: string;
  activityType: ActivityType;
  vintageYear: number;
  periodStart: string;
  periodEnd: string;
  declaredFigures: Record<string, string>;
  requestedAmount: string;
  computedCeiling: string;
  capacityBasis: string;
  capacitySource: string;
  discountFactor: string;
  referenceFactorValue: string;
  referenceFactorId: string;
  referenceLookupKey: string;
  status: ClaimStatus;
  priority: ClaimPriority;
  requiresDualApproval: boolean;
  exclusivityAttestedAt: string;
  issuedAmount: string | null;
  createdAt: string;
  evidence: ClaimEvidenceRecord[];
};

export type ClaimPage = {
  claims: ClaimRecord[];
  nextCursor: string | null;
};

export type CeilingPreview = {
  ceiling: string;
  vintageCeiling: string;
  consumed: string;
  remaining: string;
  periodCeiling: string;
  capacityBasis: string;
  capacitySource: string;
  discountFactor: string;
  referenceValue: string;
  gridRegion: string;
  periodDays: number;
  vintageDays: number;
};

export type ClaimDraft = {
  facilityId: string;
  activityType: ActivityType;
  vintageYear: number;
  periodStart: string;
  periodEnd: string;
  declaredFigures: Record<string, string>;
  requestedAmount: string;
  evidenceIds: string[];
  exclusivityAttested: boolean;
};

export type CeilingRequest = {
  facilityId: string;
  activityType: ActivityType;
  vintageYear: number;
  periodStart: string;
  periodEnd: string;
};

const toEvidence = (attachment: ApiClaimEvidence): ClaimEvidenceRecord => ({
  evidenceId: attachment.evidence_id,
  fileName: attachment.file_name,
  mediaType: attachment.media_type,
  contentHash: attachment.content_hash,
  pageCount: attachment.page_count,
  byteSize: attachment.byte_size,
});

const toClaim = (claim: ApiClaim): ClaimRecord => ({
  id: claim.id,
  facilityId: claim.facility_id,
  facilityName: claim.facility_name,
  activityType: claim.activity_type,
  vintageYear: claim.vintage_year,
  periodStart: claim.period_start,
  periodEnd: claim.period_end,
  declaredFigures: claim.declared_figures ?? {},
  requestedAmount: claim.requested_amount,
  computedCeiling: claim.computed_ceiling,
  capacityBasis: claim.capacity_basis,
  capacitySource: claim.capacity_source,
  discountFactor: claim.discount_factor,
  referenceFactorValue: claim.reference_factor_value,
  referenceFactorId: claim.reference_factor_id,
  referenceLookupKey: claim.reference_lookup_key,
  status: claimStatusNames[claim.status],
  priority: claim.priority,
  requiresDualApproval: claim.requires_dual_approval,
  exclusivityAttestedAt: claim.exclusivity_attested_at,
  issuedAmount: claim.issued_amount,
  createdAt: claim.created_at,
  evidence: (claim.evidence ?? []).map(toEvidence),
});

export const fetchClaims = async (
  token: string,
  status?: ClaimStatus,
): Promise<ClaimPage> => {
  const wanted = status ? apiClaimStatuses[status] : undefined;
  const query = wanted ? `?status=${wanted}` : "";
  const listed = await gatewayGet<{
    claims: ApiClaim[];
    next_cursor: string | null;
  }>(`/v1/claims${query}`, token);

  return {
    claims: listed.claims.map(toClaim),
    nextCursor: listed.next_cursor,
  };
};

export const fetchClaim = async (
  token: string,
  claimId: string,
): Promise<ClaimRecord> => {
  const found = await gatewayGet<ApiClaim>(`/v1/claims/${claimId}`, token);
  return toClaim(found);
};

export const submitClaim = async (
  token: string,
  draft: ClaimDraft,
  idempotencyKey: string,
): Promise<ClaimRecord> => {
  const created = await gatewayPost<ApiClaim>(
    "/v1/claims",
    token,
    {
      facility_id: draft.facilityId,
      activity_type: draft.activityType,
      vintage_year: draft.vintageYear,
      period_start: draft.periodStart,
      period_end: draft.periodEnd,
      declared_figures: draft.declaredFigures,
      requested_amount: draft.requestedAmount,
      evidence_ids: draft.evidenceIds,
      exclusivity_attested: draft.exclusivityAttested,
    },
    idempotencyKey,
  );

  return toClaim(created);
};

export const previewCeiling = async (
  token: string,
  request: CeilingRequest,
  idempotencyKey: string,
): Promise<CeilingPreview> => {
  const preview = await gatewayPost<{
    ceiling: string;
    vintage_ceiling: string;
    consumed: string;
    remaining: string;
    period_ceiling: string;
    capacity_basis: string;
    capacity_source: string;
    discount_factor: string;
    reference_value: string;
    grid_region: string;
    period_days: number;
    vintage_days: number;
  }>(
    "/v1/claims/ceiling-preview",
    token,
    {
      facility_id: request.facilityId,
      activity_type: request.activityType,
      vintage_year: request.vintageYear,
      period_start: request.periodStart,
      period_end: request.periodEnd,
    },
    idempotencyKey,
  );

  return {
    ceiling: preview.ceiling,
    vintageCeiling: preview.vintage_ceiling,
    consumed: preview.consumed,
    remaining: preview.remaining,
    periodCeiling: preview.period_ceiling,
    capacityBasis: preview.capacity_basis,
    capacitySource: preview.capacity_source,
    discountFactor: preview.discount_factor,
    referenceValue: preview.reference_value,
    gridRegion: preview.grid_region,
    periodDays: preview.period_days,
    vintageDays: preview.vintage_days,
  };
};

File: backend/services/api-gateway/internal/handler/claims.go
package handler

import (
	"net/http"
	"strconv"

	"github.com/gin-gonic/gin"
	"google.golang.org/genproto/googleapis/rpc/errdetails"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"

	sustainabilityv1 "github.com/carboncircuit/backend/gen/carboncircuit/sustainability/v1"
	"github.com/carboncircuit/backend/internal/events"
	"github.com/carboncircuit/backend/internal/httpx"
	"github.com/carboncircuit/backend/services/api-gateway/internal/caller"
)

var activityTypeByName = map[string]sustainabilityv1.ActivityType{
	"renewable_energy":           sustainabilityv1.ActivityType_ACTIVITY_TYPE_RENEWABLE_ENERGY,
	"reduced_emission_logistics": sustainabilityv1.ActivityType_ACTIVITY_TYPE_REDUCED_EMISSION_LOGISTICS,
	"responsible_sourcing":       sustainabilityv1.ActivityType_ACTIVITY_TYPE_RESPONSIBLE_SOURCING,
}

var activityTypeName = invertMap(activityTypeByName)

var claimStatusByName = map[string]sustainabilityv1.ClaimStatus{
	"submitted":                  sustainabilityv1.ClaimStatus_CLAIM_STATUS_SUBMITTED,
	"ai_review":                  sustainabilityv1.ClaimStatus_CLAIM_STATUS_AI_REVIEW,
	"human_review":               sustainabilityv1.ClaimStatus_CLAIM_STATUS_HUMAN_REVIEW,
	"approved":                   sustainabilityv1.ClaimStatus_CLAIM_STATUS_APPROVED,
	"rejected":                   sustainabilityv1.ClaimStatus_CLAIM_STATUS_REJECTED,
	"more_information_requested": sustainabilityv1.ClaimStatus_CLAIM_STATUS_MORE_INFORMATION_REQUESTED,
}

var claimStatusName = invertMap(claimStatusByName)

var priorityName = map[sustainabilityv1.QueuePriority]string{
	sustainabilityv1.QueuePriority_QUEUE_PRIORITY_NORMAL:   "normal",
	sustainabilityv1.QueuePriority_QUEUE_PRIORITY_HIGH:     "high",
	sustainabilityv1.QueuePriority_QUEUE_PRIORITY_CRITICAL: "critical",
}

type submitClaimRequest struct {
	FacilityID          string            `json:"facility_id" binding:"required,uuid"`
	ActivityType        string            `json:"activity_type" binding:"required"`
	VintageYear         int32             `json:"vintage_year" binding:"required"`
	PeriodStart         string            `json:"period_start" binding:"required"`
	PeriodEnd           string            `json:"period_end" binding:"required"`
	DeclaredFigures     map[string]string `json:"declared_figures"`
	RequestedAmount     string            `json:"requested_amount" binding:"required,max=32"`
	EvidenceIDs         []string          `json:"evidence_ids" binding:"required,min=1,max=25,dive,uuid"`
	ExclusivityAttested bool              `json:"exclusivity_attested"`
}

type previewCeilingRequest struct {
	FacilityID   string `json:"facility_id" binding:"required,uuid"`
	ActivityType string `json:"activity_type" binding:"required"`
	VintageYear  int32  `json:"vintage_year" binding:"required"`
	PeriodStart  string `json:"period_start" binding:"required"`
	PeriodEnd    string `json:"period_end" binding:"required"`
}

type claimEvidenceResponse struct {
	EvidenceID  string `json:"evidence_id"`
	FileName    string `json:"file_name"`
	MediaType   string `json:"media_type"`
	ContentHash string `json:"content_hash"`
	PageCount   *int32 `json:"page_count"`
	ByteSize    int64  `json:"byte_size"`
}

type claimResponse struct {
	ID                    string                  `json:"id"`
	FacilityID            string                  `json:"facility_id"`
	FacilityName          string                  `json:"facility_name"`
	ActivityType          string                  `json:"activity_type"`
	VintageYear           int32                   `json:"vintage_year"`
	PeriodStart           string                  `json:"period_start"`
	PeriodEnd             string                  `json:"period_end"`
	DeclaredFigures       map[string]string       `json:"declared_figures"`
	RequestedAmount       string                  `json:"requested_amount"`
	ComputedCeiling       string                  `json:"computed_ceiling"`
	CapacityBasis         string                  `json:"capacity_basis"`
	CapacitySource        string                  `json:"capacity_source"`
	DiscountFactor        string                  `json:"discount_factor"`
	ReferenceFactorValue  string                  `json:"reference_factor_value"`
	ReferenceFactorID     string                  `json:"reference_factor_id"`
	ReferenceLookupKey    string                  `json:"reference_lookup_key"`
	Status                string                  `json:"status"`
	Priority              string                  `json:"priority"`
	RequiresDualApproval  bool                    `json:"requires_dual_approval"`
	ExclusivityAttestedAt string                  `json:"exclusivity_attested_at"`
	IssuedAmount          *string                 `json:"issued_amount"`
	CreatedAt             string                  `json:"created_at"`
	Evidence              []claimEvidenceResponse `json:"evidence"`
}

func toClaimResponse(claim *sustainabilityv1.Claim) claimResponse {
	evidence := make([]claimEvidenceResponse, 0, len(claim.GetEvidence()))
	for _, attachment := range claim.GetEvidence() {
		var pages *int32
		if attachment.GetPageCount() > 0 {
			counted := attachment.GetPageCount()
			pages = &counted
		}
		evidence = append(evidence, claimEvidenceResponse{
			EvidenceID:  attachment.GetEvidenceId(),
			FileName:    attachment.GetFileName(),
			MediaType:   attachment.GetMediaType(),
			ContentHash: attachment.GetContentHash(),
			PageCount:   pages,
			ByteSize:    attachment.GetByteSize(),
		})
	}

	return claimResponse{
		ID:                    claim.GetId(),
		FacilityID:            claim.GetFacilityId(),
		FacilityName:          claim.GetFacilityName(),
		ActivityType:          activityTypeName[claim.GetActivityType()],
		VintageYear:           claim.GetVintageYear(),
		PeriodStart:           claim.GetPeriodStart(),
		PeriodEnd:             claim.GetPeriodEnd(),
		DeclaredFigures:       claim.GetDeclaredFigures(),
		RequestedAmount:       claim.GetRequestedAmount(),
		ComputedCeiling:       claim.GetComputedCeiling(),
		CapacityBasis:         claim.GetCapacityBasis(),
		CapacitySource:        claim.GetCapacitySource(),
		DiscountFactor:        claim.GetDiscountFactor(),
		ReferenceFactorValue:  claim.GetReferenceFactorValue(),
		ReferenceFactorID:     claim.GetReferenceFactorId(),
		ReferenceLookupKey:    claim.GetReferenceLookupKey(),
		Status:                claimStatusName[claim.GetStatus()],
		Priority:              priorityName[claim.GetPriority()],
		RequiresDualApproval:  claim.GetRequiresDualApproval(),
		ExclusivityAttestedAt: claim.GetExclusivityAttestedAt(),
		IssuedAmount:          emptyToNil(claim.GetIssuedAmount()),
		CreatedAt:             claim.GetCreatedAt(),
		Evidence:              evidence,
	}
}

func (h *Handlers) SubmitClaim(c *gin.Context) {
	if _, resolved := caller.ContextFrom(c.Request.Context()); !resolved {
		httpx.Fail(c, httpx.CodeUnauthenticated)
		return
	}

	idempotencyKey, present := httpx.IdempotencyKeyFrom(c.Request.Context())
	if !present {
		httpx.Fail(c, httpx.CodeIdempotencyKeyRequired)
		return
	}

	var body submitClaimRequest
	if err := c.ShouldBindJSON(&body); err != nil {
		httpx.Fail(c, httpx.CodeValidation)
		return
	}

	activity, known := activityTypeByName[body.ActivityType]
	if !known {
		httpx.Fail(c, httpx.CodeValidation, httpx.FieldError{
			Field: "activity_type", Code: "UNSUPPORTED_VALUE",
		})
		return
	}

	if !body.ExclusivityAttested {
		httpx.Fail(c, httpx.CodeValidation, httpx.FieldError{
			Field: "exclusivity_attested", Code: "REQUIRED",
		})
		return
	}

	submitted, err := h.Sustainability.SubmitClaim(c.Request.Context(), idempotencyKey,
		&sustainabilityv1.SubmitClaimRequest{
			FacilityId:          body.FacilityID,
			ActivityType:        activity,
			VintageYear:         body.VintageYear,
			PeriodStart:         body.PeriodStart,
			PeriodEnd:           body.PeriodEnd,
			DeclaredFigures:     body.DeclaredFigures,
			RequestedAmount:     body.RequestedAmount,
			EvidenceIds:         body.EvidenceIDs,
			ExclusivityAttested: body.ExclusivityAttested,
			IdempotencyKey:      idempotencyKey,
		})
	if err != nil {
		h.failClaim(c, err)
		return
	}

	statusCode := http.StatusCreated
	if submitted.GetAlreadySubmitted() {
		statusCode = http.StatusOK
	}

	httpx.Data(c, statusCode, toClaimResponse(submitted.GetClaim()))
}

func (h *Handlers) ListClaims(c *gin.Context) {
	if _, resolved := caller.ContextFrom(c.Request.Context()); !resolved {
		httpx.Fail(c, httpx.CodeUnauthenticated)
		return
	}

	limit := 0
	if raw := c.Query("limit"); raw != "" {
		parsed, err := strconv.Atoi(raw)
		if err != nil || parsed < 1 {
			httpx.Fail(c, httpx.CodeValidation, httpx.FieldError{
				Field: "limit", Code: "UNSUPPORTED_VALUE",
			})
			return
		}
		limit = parsed
	}

	listed, err := h.Sustainability.ListClaims(c.Request.Context(),
		&sustainabilityv1.ListClaimsRequest{
			Status: claimStatusByName[c.Query("status")],
			After:  c.Query("after"),
			Limit:  int32(limit),
		})
	if err != nil {
		h.failClaim(c, err)
		return
	}

	claims := make([]claimResponse, 0, len(listed.GetClaims()))
	for _, claim := range listed.GetClaims() {
		claims = append(claims, toClaimResponse(claim))
	}

	httpx.Data(c, http.StatusOK, map[string]any{
		"claims":      claims,
		"next_cursor": emptyToNil(listed.GetNextCursor()),
	})
}

func (h *Handlers) GetClaim(c *gin.Context) {
	if _, resolved := caller.ContextFrom(c.Request.Context()); !resolved {
		httpx.Fail(c, httpx.CodeUnauthenticated)
		return
	}

	found, err := h.Sustainability.GetClaim(c.Request.Context(), c.Param("claimId"))
	if err != nil {
		h.failClaim(c, err)
		return
	}

	httpx.Data(c, http.StatusOK, toClaimResponse(found.GetClaim()))
}

func (h *Handlers) PreviewClaimCeiling(c *gin.Context) {
	if _, resolved := caller.ContextFrom(c.Request.Context()); !resolved {
		httpx.Fail(c, httpx.CodeUnauthenticated)
		return
	}

	var body previewCeilingRequest
	if err := c.ShouldBindJSON(&body); err != nil {
		httpx.Fail(c, httpx.CodeValidation)
		return
	}

	activity, known := activityTypeByName[body.ActivityType]
	if !known {
		httpx.Fail(c, httpx.CodeValidation, httpx.FieldError{
			Field: "activity_type", Code: "UNSUPPORTED_VALUE",
		})
		return
	}

	preview, err := h.Sustainability.PreviewCeiling(c.Request.Context(),
		&sustainabilityv1.PreviewCeilingRequest{
			FacilityId:   body.FacilityID,
			ActivityType: activity,
			VintageYear:  body.VintageYear,
			PeriodStart:  body.PeriodStart,
			PeriodEnd:    body.PeriodEnd,
		})
	if err != nil {
		h.failClaim(c, err)
		return
	}

	httpx.Data(c, http.StatusOK, map[string]any{
		"ceiling":         preview.GetCeiling(),
		"capacity_basis":  preview.GetCapacityBasis(),
		"capacity_source": preview.GetCapacitySource(),
		"discount_factor": preview.GetDiscountFactor(),
		"reference_value": preview.GetReferenceValue(),
		"grid_region":     preview.GetGridRegion(),
		"period_days":     preview.GetPeriodDays(),
		"vintage_days":    preview.GetVintageDays(),
	})
}

func claimRefusalReason(err error) string {
	for _, detail := range status.Convert(err).Details() {
		info, matches := detail.(*errdetails.ErrorInfo)
		if matches && info.GetDomain() == events.ClaimRefusalDomain {
			return info.GetReason()
		}
	}
	return "REFUSED"
}

func (h *Handlers) failClaim(c *gin.Context, err error) {
	reported := status.Convert(err)

	switch reported.Code() {
	case codes.NotFound:
		httpx.Fail(c, httpx.CodeResourceNotFound)
	case codes.PermissionDenied:
		httpx.Fail(c, httpx.CodeOrganizationReadOnly)
	case codes.FailedPrecondition:
		httpx.Fail(c, httpx.CodeValidation, httpx.FieldError{
			Field: "claim", Code: claimRefusalReason(err),
		})
	case codes.InvalidArgument:
		httpx.Fail(c, httpx.CodeValidation)
	case codes.AlreadyExists:
		httpx.Fail(c, httpx.CodeIdempotencyKeyReused)
	case codes.Aborted:
		httpx.Fail(c, httpx.CodeRequestInProgress)
	default:
		h.Logger.Error("claim call failed", errorAttributes(c, err)...)
		httpx.Fail(c, httpx.CodeDependencyUnavailable)
	}
}
